src/Maestro/Maestro.Web/Api/v2019_01_16/Controllers/BuildTimeController.cs

```diff
@@ -0,0 +1,128 @@
+        {
+            Data.Models.DefaultChannel defaultChannel = await _context.DefaultChannels.FindAsync(id);
+
+            if (defaultChannel == null)
+            {
+                return NotFound();
+            }
+
+            var properties = new ClientRequestProperties();
+
+            // There is a mismatch between how we store the names of the repositories in BAR
+            // and how we store them in Kusto. Normalize to the last part of the url.
+            properties.SetParameter("_Repository", defaultChannel.Repository.Split("/").Last());
+            properties.SetParameter("_SourceBranch", defaultChannel.Branch);
+
+            string queryParameters = String.Join(",", properties.Parameters.Select(p => $"{p.Key}:{KustoDataTypes.String}"));
+
+            // We only care about builds that complete successfully or partially successfully 
+            // from the given repository. We summarize duration of the builds over the last 7 days.
+            // There are multiple different definitions that run in parallel, so we summarize
+            // on the definition id and ultimately choose the definition that took the longest.
+            string commonQueryText = @"| where Repository endswith _Repository
+                | where Result != 'failed' and Result != 'canceled' 
+                | where FinishTime > ago(7d) 
+                | extend duration = FinishTime - StartTime 
+                | summarize avg(duration) by DefinitionId
+                | summarize max(avg_duration)";
+
+            // We only want the pull request time from the public ci. We don't exclude on branch,
+            // as all PRs come in as refs/heads/#/merge rather than what branch they are trying to
+            // apply to.
+            string publicQuery = $@"TimelineBuilds 
+                | where Project == 'public'
+                | where Reason == 'pullRequest' 
+                {commonQueryText}";
+
+            // For the official build times, we want the builds that were generated as a CI run 
+            // (either batchedCI or individualCI) for a specific branch--i.e. we want the builds
+            // that are part of generating the product.
+            string internalQuery = $@"TimelineBuilds 
+                | where Project == 'internal' 
+                | where Reason contains 'CI'
+                | where SourceBranch endswith _SourceBranch
+                {commonQueryText}";
+                
+            double officialTime = 0;
+            double prTime = 0;
+
+            using (ICslQueryProvider query =
+                    _kustoClientProvider.GetKustoQueryConnectionProvider())
+            using (IDataReader result = await query.ExecuteQueryAsync(
+                _kustoClientProvider.GetKustoDatabase(),
+                $"declare query_parameters ({queryParameters});{publicQuery}",
+                properties
+            ))
+            {
+                while (result.Read())
```

Comment:
Why loop over a bunch if we are only going to take the last value? The query to should be changed to only return a singe value, and then we should do `if (reader.Read()) { use it}; if (reader.Read()) { throw some excpetion or log an error because we didn't expect a second row}`

Reply:
Or, even better, you could make a helper method that is "read exactly one"

Reply:
This is basically a remnant of my initial implementation where I had a list.